src/nupic/regions/SPRegion.cpp

```diff
@@ -0,0 +1,1266 @@
+  // concatination of all incomming buffers.
+  Array &inputBuffer = getInput("bottomUpIn")->getData();
+  args_.inputWidth = (UInt32)inputBuffer.getCount();
+  if (args_.inputWidth == 0) {
+    NTA_THROW << "SPRegion::initialize - No input was provided.\n";
+  }
+
+  std::vector<UInt32> inputDimensions = {args_.inputWidth};
+  std::vector<UInt32> columnDimensions = {columnCount};
+  if (args_.potentialRadius == 0)
+    args_.potentialRadius = args_.inputWidth;
+
+  sp_ = new SpatialPooler(
+      inputDimensions, columnDimensions, args_.potentialRadius,
+      args_.potentialPct, args_.globalInhibition, args_.localAreaDensity,
+      args_.numActiveColumnsPerInhArea, args_.stimulusThreshold,
+      args_.synPermInactiveDec, args_.synPermActiveInc, args_.synPermConnected,
+      args_.minPctOverlapDutyCycles, args_.dutyCyclePeriod, args_.boostStrength,
+      args_.seed, args_.spVerbosity, args_.wrapAround);
+}
+
+void SPRegion::compute() {
+  // Note: the Python code has a hook at this point to activate profiling with
+  // hotshot.  This version does not provide this hook although there are
+  // several C++ profilers that could be used.
+
+  NTA_ASSERT(sp_) << "SP not initialized";
+
+  if (args_.topDownMode) {
+    // TOP-DOWN inference mode
+    NTA_THROW << "Top Down Inference mode is not implemented.";
+  } else {
+    // BOTTOM-UP compute mode
+    args_.iter++;
+
+    // Note: The Input and Output objects are UInt32 types (containing 1's and
+    // 0's).
+    nzInputValid_ = false;
+
+    /**************************************************  Not used.
+    // check for reset
+    bool resetSignal = false;
+    Input* resetIn = region_->getInput("resetIn");
+    if (resetIn)
+    {
+      // if there is a resetIn array and its first element is not 0 then the
+    resetSignal is true. const Array& resetArray = resetIn->getData();
+      NTA_CHECK(resetArray.getCount() == 1);
+      resetSignal = (*((Real32*)resetArray.getBuffer()) != 0.0);
+    }
+    ***************************************************/
+
+    // Call SpatialPooler
+    Array &inputBuffer = getInput("bottomUpIn")->getData();
+    Array &outputBuffer = getOutput("bottomUpOut")->getData();
+
+    UInt32 *inputVector = (UInt32 *)inputBuffer.getBuffer();
+    UInt32 *outputVector = (UInt32 *)outputBuffer.getBuffer();
+    sp_->compute(inputVector, args_.learningMode, outputVector);
+
+    // Prepare the output
+    nzOutputValid_ = false;
+
+    size_t size = sp_->getNumColumns();
+    // Direct logging of SP non-zero outputs if requested
+    if (!logPathOutput_.empty()) {
+      FILE *fp = fopen(logPathOutput_.c_str(), "a");
+      if (fp) {
+        for (int i = 0; i < size; i++) {
+          if (outputVector[i])
+            fprintf(fp, "%d ", i);
+        }
+        fprintf(fp, "\n");
+        fclose(fp);
+      }
+    }
+
+    // Direct logging of SP non-zero inputs if requested
+    if (!logPathInput_.empty()) {
+      FILE *fp = fopen(logPathInput_.c_str(), "a");
```

Comment:
This is basically a direct port from the .py code.  So what you see are things that the .py supports.

The file open could be replaced with a stream I guess.

Reply:
> The file open could be replaced with a stream I guess. 

I would even go and remove this logging possibility (?)

Reply:
For TM's I commented out the logging and the parameters for it. (keeping the code just in case it is important to someone). You could do the same for SP.

Reply:
commended out